### PR TITLE
Add 'edit' wishlist functionality

### LINKED ISSUE #13: [Dependent on Issues #8, #6] Add Rule - Only Signed in user can edit a wishlist
URI /wishlists/:id/edit
wishlists#edit

### LINKED ISSUE #27: [Dependent on Issue #24] Updating specific wishlist
Separate options for changing wishlist name, description, venue, date, particular wishlist item (with its name, link to the product, info about gift buyer).

### PR DESCRIPTION
**Pull request by agalkiewicz**
Closes #13, closes #27.

### CODE REVIEW

File: app/controllers/wishlists_controller.rb [hunk 1]
```diff
@@ -30,6 +30,25 @@ def destroy
     redirect_to wishlists_path, alert: t("generic_error")
   end
 
+  def edit
+    @wishlist = Wishlist.find(params[:id])
```
Comment (smt116) [inline app/controllers/wishlists_controller.rb]: this is exactly the same instruction like in other actions. You can extract it to `before_action`

File: app/models/ability.rb [hunk 2]
```diff
@@ -3,7 +3,7 @@ class Ability
 
   def initialize(user)
     if user
-      can [:create, :read, :destroy], Wishlist, user_id: user.id
+      can %i(create read destroy update), Wishlist, user_id: user.id
```
Comment (smt116) [inline app/models/ability.rb:6]: 👍

File: spec/controllers/wishlists_controller_spec.rb [hunk 3]
```diff
@@ -100,15 +101,86 @@
...
+
+        it "renders the 'edit' template" do
+          expect(response).to render_template(:edit)
+        end
+
+        context "and doesn't belong to logged in user" do
+          let!(:alice) { create(:alice) }
+          let!(:wishlist) { create(:wishlist, user: alice) }
+
+          subject! { get_edit(id: wishlist.id) }
```
Comment (smt116) [inline spec/controllers/wishlists_controller_spec.rb]: no need to redefine it here. subject from line 109 should work as well

File: spec/controllers/wishlists_controller_spec.rb [hunk 4]
```diff
@@ -100,15 +101,86 @@
...
+        end
 
         it "shows an error" do
           expect(flash[:alert]).not_to be_nil
         end
       end
     end
+
+    describe "PUT #update" do
+      let!(:new_params) { { name: "name" } }
```
Comment (smt116) [inline spec/controllers/wishlists_controller_spec.rb]: why is it `let!` instead of `let`?

File: spec/controllers/wishlists_controller_spec.rb [hunk 5]
```diff
@@ -144,6 +216,34 @@
...
+        before(:each) { get_edit(id: wishlist_id) }
+
+        include_examples "redirects to the 'sign_in' page"
+      end
+
+      context "with invalid parameters" do
+        before(:each) { post_create(id: -1) }
+
+        include_examples "redirects to the 'sign_in' page"
+      end
```
Comment (smt116) [inline spec/controllers/wishlists_controller_spec.rb:211]: you have duplication here. I would put a context for not logged in user to the previous describe "GET #edit".

File: spec/controllers/wishlists_controller_spec.rb [hunk 6]
```diff
@@ -144,6 +216,34 @@
...
+
+    describe "GET #edit" do
+      context "with valid parameters" do
+        before(:each) { get_edit(id: wishlist_id) }
+
+        include_examples "redirects to the 'sign_in' page"
+      end
+
+      context "with invalid parameters" do
+        before(:each) { post_create(id: -1) }
```
Comment (smt116) [inline spec/controllers/wishlists_controller_spec.rb]: post_create?

Comment (agalkiewicz) [inline spec/controllers/wishlists_controller_spec.rb]: Oh God, what a shame. :P

Comment (agalkiewicz) [inline spec/controllers/wishlists_controller_spec.rb]: Too many ctrl + z. :P

File: spec/controllers/wishlists_controller_spec.rb (hunk 4 above)
Comment (agalkiewicz) [inline spec/controllers/wishlists_controller_spec.rb]: I don't really know why I have used it. Must be more careful.

File: app/views/wishlists/edit.html.erb [hunk 7]
```diff
@@ -0,0 +1,3 @@
+<h1>Update your wishlist!</h1>
+<% provide(:button_name, "Update") %>
+<%= render partial: "wishlist_form" %>
```
Comment (student-tomasz) [inline app/views/wishlists/edit.html.erb]: Why not pass the button's name as a local?

File: app/views/wishlists/index.html.erb [hunk 8]
```diff
@@ -3,12 +3,13 @@
   <ul>
     <% @wishlists.each do |wishlist| %>
       <li>
-        <%= link_to wishlist.name, "#" -%>
-        <%= button_to "Delete", wishlist, method: :delete -%>
+        <%= link_to wishlist.name, "#" %>
+        <%= button_to "Edit", edit_wishlist_path(wishlist), method: :get %>
```
Comment (student-tomasz) [inline app/views/wishlists/index.html.erb]: `link_to` is enough. With `button_to` you're sending a POST request from a form, that later gets changed to a GET.

File: config/routes.rb [hunk 9]
```diff
@@ -1,6 +1,6 @@
 Rails.application.routes.draw do
   root to: "wishlists#index"
-  resources :wishlists, only: [:index, :new, :create, :destroy]
+  resources :wishlists, only: %i(index new create destroy edit update)
```
Comment (student-tomasz) [inline config/routes.rb:3]: 😢

File: spec/controllers/wishlists_controller_spec.rb [hunk 10]
```diff
@@ -3,6 +3,7 @@
 RSpec.describe WishlistsController do
   context "when user is logged in" do
     before(:each) { sign_in_user }
+    let!(:wishlist) { create(:wishlist, user: controller.current_user) }
```
Comment (student-tomasz) [inline spec/controllers/wishlists_controller_spec.rb]: So, why are you using the `let!` variant instead of `let`?
And does it really need to be all the way up here? It's not need in `#index`, `#new` and `#create` tests, and duplicates inside `#destroy` specs.

File: spec/controllers/wishlists_controller_spec.rb [hunk 11]
```diff
@@ -100,15 +101,84 @@
...
+    describe "GET #edit" do
+      context "when wishlist exists" do
+        subject! { get_edit(id: wishlist.id) }
+
+        it "returns 200 OK" do
+          expect(response).to have_http_status(:ok)
+        end
+
+        it "renders the 'edit' template" do
+          expect(response).to render_template(:edit)
```
Comment (student-tomasz) [inline spec/controllers/wishlists_controller_spec.rb:115]: Why do you use different syntaxes in different tests? You can use `is_expected` here, or `expect(response)` below, why change?

File: spec/controllers/wishlists_controller_spec.rb [hunk 12]
```diff
@@ -100,15 +101,84 @@
...
       end
     end
+
+    describe "PUT #update" do
+      let(:new_params) { { name: "name" } }
+
+      context "when params valid" do
+        subject! { put_update(wishlist.id, new_params) }
+
+        it "changes object attributes" do
```
Comment (student-tomasz) [inline spec/controllers/wishlists_controller_spec.rb]: It's the wishlist, not some ambiguous object.

File: spec/controllers/wishlists_controller_spec.rb (hunk 10 above)
Comment (agalkiewicz) [inline spec/controllers/wishlists_controller_spec.rb]: It looks like I've simply followed Your approach (delete action). But of course it isn't an excuse and both "!" and duplication in 'destroy' aren't necessary. Do You think defining wishlist separately in 'delete', 'update' and 'get' is a good option? I prefer to put it here.

File: spec/controllers/wishlists_controller_spec.rb (hunk 11 above)
Comment (agalkiewicz) [inline spec/controllers/wishlists_controller_spec.rb]: There is so many options that I cannot decide which to choose. :P

Comment (agalkiewicz) [inline spec/controllers/wishlists_controller_spec.rb:115]: The same question for You (post, delete). :) Maybe the sameness is a good option but I think that it's not too important and good to know that there are many instructions of the same purpose. What's important - the same instruction is used in a whole block of tests describing one specific method.

File: app/controllers/wishlists_controller.rb [hunk 13]
```diff
@@ -1,5 +1,6 @@
 class WishlistsController < ApplicationController
   before_action :authenticate_user!
+  before_action :find_wishlist, only: %i(destroy edit update)
```
Comment (smt116) [inline app/controllers/wishlists_controller.rb]: `set_wishlist` by convention

File: app/controllers/wishlists_controller.rb [hunk 14]
```diff
@@ -37,4 +54,10 @@ def wishlist_params
     params[:wishlist][:user_id] = current_user.id
     params.require(:wishlist).permit(:user_id, :name, :description)
   end
+
+  def find_wishlist
+    @wishlist = Wishlist.find(params[:id])
+  rescue ActiveRecord::RecordNotFound
+    redirect_to wishlists_path, alert: t("generic_error")
```
Comment (smt116) [inline app/controllers/wishlists_controller.rb:60]: It is okay for now, but generally it should be handled somewhere on the higher level to catch all those exceptions and render error page.

Comment (smt116) [inline app/controllers/wishlists_controller.rb:60]: and in this case method name could be even `set_wishlist!` to emphasize that exception will be raised

File: config/routes.rb (hunk 9 above)
Comment (smt116) [inline config/routes.rb:3]: imho good change 👍

File: spec/controllers/wishlists_controller_spec.rb (hunk 11 above)
Comment (smt116) [inline spec/controllers/wishlists_controller_spec.rb:115]: use `is_expected` with single line specs (`it { }`) and use `expect(foo)` with block (`it "..." do ...`). Here it is okay to use `expect(response)`

File: spec/controllers/wishlists_controller_spec.rb [hunk 15]
```diff
@@ -100,20 +100,72 @@
         it "redirects to the wishlists index page" do
           is_expected.to redirect_to(wishlists_path)
         end
+      end
+    end
+
+    describe "GET #edit" do
+      context "when wishlist exists" do
+        subject! { get_edit(id: wishlist.id) }
```
Comment (smt116) [inline spec/controllers/wishlists_controller_spec.rb]: it should be `before` block instead. You are not using `subject` in assertion below.

File: spec/controllers/wishlists_controller_spec.rb (hunk 11 above)
Comment (student-tomasz) [inline spec/controllers/wishlists_controller_spec.rb:115]: @agalkiewicz Yep, no one called me out on it, and I didn't even notice. Change the inconsistencies there as well, ok?

Comment (agalkiewicz) [inline spec/controllers/wishlists_controller_spec.rb:115]: :+1:

File: spec/controllers/wishlists_controller_spec.rb (hunk 5 above)
Comment (agalkiewicz) [inline spec/controllers/wishlists_controller_spec.rb:211]: Have thought about this change but for me it's good now. Dividing tests into 2 main contexts "when user logged in" and vice versa seems fine for me. Or maybe I haven't understand You correctly.

Comment (smt116) [inline spec/controllers/wishlists_controller_spec.rb:211]: For me it should be like this:

```
describe #edit
  context when user is logged in
    context with valid params
      assertions
    context with invalid params
      assertion
  context when user is logged out
    assertions for valid params # imho there is no need to check invalid here
```

File: config/routes.rb (hunk 9 above)
Comment (student-tomasz) [inline config/routes.rb:3]: Yeah, I'm not complaining. I knew it was inevitable. Just a reference to our previous conversation about this precise line. 😜

File: spec/controllers/wishlists_controller_spec.rb (hunk 5 above)
Comment (agalkiewicz) [inline spec/controllers/wishlists_controller_spec.rb:211]: I kept check of invalid params when user is logged out to show that the action is the same as when params are valid (that sign_up page is displayed, not sth like error 404). But maybe it's unnecessary indeed. I'll try to change schema of tests as You said.

File: app/controllers/wishlists_controller.rb [hunk 16]
```diff
@@ -22,11 +23,27 @@ def create
...
+
+  def edit
+    authorize! :edit, @wishlist
+  rescue CanCan::AccessDenied
+    redirect_to wishlists_path, alert: t("generic_error")
+  end
+
+  def update
+    authorize! :update, @wishlist
+    if @wishlist.update_attributes(wishlist_params)
```
Comment (student-tomasz) [inline app/controllers/wishlists_controller.rb]: _That's just a note._ You could use `update!` to avoid using `if`, and then handle the bad path in a `rescue` block catching `ActiveRecord::RecordInvalid`. But that's just my personal habit, I tend to avoid using `if` as much as possible.

File: spec/controllers/wishlists_controller_spec.rb (hunk 5 above)
Comment (agalkiewicz) [inline spec/controllers/wishlists_controller_spec.rb:211]: While I was changing the schema, I thought that it's not worth it. Duplications of other lines occur. Can I leave it alone and only delete checking invalid params, when user's logged out, or duplication of `describe #method` is a bad thing?

File: app/controllers/wishlists_controller.rb (hunk 16 above)
Comment (agalkiewicz) [inline app/controllers/wishlists_controller.rb]: Nice one, thanks. :)

**Comment by gotar** [comment]
rebase (squash) and it can go
